source/slang/source-loc.cpp

```diff
@@ -70,118 +301,64 @@ SourceRange SourceManager::allocateSourceRange(UInt size)
-    String const&       path,
-    UInt                line)
+SourceUnit* SourceManager::newSourceUnit(SourceFile* sourceFile)
 {
-    // First, we need to find out what file we are being asked to remap
-    ExpandedSourceLoc expandedDirectiveLoc = expandSourceLoc(getSpellingLoc(directiveLoc));
-    HumaneSourceLoc humaneDirectiveLoc = getHumaneLoc(expandedDirectiveLoc);
+    SourceRange range = allocateSourceRange(sourceFile->content.size());
+    SourceUnit* sourceUnit = new SourceUnit(this, sourceFile, range);
+    m_sourceUnits.Add(sourceUnit);
 
-    SourceFile* sourceFile = expandedDirectiveLoc.getSourceFile();
-    if(!sourceFile)
-        return SourceLoc();
-
-    // We are going to be wasteful here and allocate a range of source locations
-    // that can cover the entire input file. This will lead to a problem with
-    // memory usage if we ever had a large input file that used many `#line` directives,
-    // since our usage of ranges would be quadratic!
-
-    // Count how many locations we'd need to reserve for a complete clone of the input
-    UInt size = sourceFile->sourceRange.end.getRaw() - sourceFile->sourceRange.begin.getRaw();
-
-    // Allocate a fresh range for our logically remapped file
-    SourceRange sourceRange = allocateSourceRange(size);
-
-    // Now fill in an entry that will point at the original source file,
-    // but use our new range.
-    Entry entry;
-    entry.sourceFile = sourceFile;
-    entry.startLoc = sourceRange.begin;
-    entry.path = path;
-
-    // We also need to make sure that any lookups for line numbers will
-    // get corrected based on this files location.
-    entry.lineAdjust = Int(line) - Int(humaneDirectiveLoc.line + 1);
-
-    sourceFiles.Add(entry);
-
-    return entry.startLoc;
+    return sourceUnit;
 }
 
-static ExpandedSourceLoc expandSourceLoc(
-    SourceManager*      inSourceManager,
-    SourceLoc const&    loc)
+SourceUnit* SourceManager::findSourceUnit(SourceLoc loc)
 {
-    SourceManager* sourceManager = inSourceManager;
-
-    ExpandedSourceLoc expanded;
-
-    SourceLoc::RawValue rawValue = loc.getRaw();
+    SourceLoc::RawValue rawLoc = loc.getRaw();
 
-    // Invalid location? -> invalid expanded location
-    if(rawValue == 0)
-        return expanded;
+    int hi = int(m_sourceUnits.Count());
 
-    // Past the end of what we can handle? -> invalid
-    if(rawValue >= sourceManager->nextLoc.getRaw())
-        return expanded;
-
-    // Maybe the location came from a parent source manager
-    while( rawValue < sourceManager->startLoc.getRaw()
-        && sourceManager->parent)
+    if (hi == 0)
     {
-        sourceManager = sourceManager->parent;
+        return nullptr;
     }
 
-    SLANG_ASSERT(sourceManager->sourceFiles.Count() > 0);
-
-    UInt lo = 0;
-    UInt hi = sourceManager->sourceFiles.Count();
-
-    while( lo+1 < hi )
+    int lo = 0;
+    while (lo + 1 < hi)
     {
-        UInt mid = lo + (hi - lo) / 2;
+        int mid = (hi + lo) >> 1;
```

Comment:
I know this will sound absolutely silly, but the original code was written the way it was on purpose. While it is unlikely to ever come up in practice, the `(hi + lo)` computation can technically overflow even when `hi` and `lo` are both in range for the `int` (or `UInt` in the original code) type, while `lo + (hi - lo)/2` can't overflow and produces the same mathematical result.

Also as a style thing, I mostly prefer just going ahead and writing `... / 2` when what you actually want is division rather than bit-twiddling. For me, using a shift obscures the intention and is unlikely to actually produce better code in practice. Just my opinion, though.